content/blog/2019-10-02-codestitcher.md

```diff
@@ -0,0 +1,166 @@
+2. Profile the program, generating weights for CFG edges.
+3. Using the weighted CFG, perform basic block chaining, ordering basic blocks in a way to minimize jumps.
+4. Perform hierarchical code collocation, ordering these basic block chains to form higher-level layouts. Here, the *d*-close partial layout algorithm works to build layouts to maximize spatial locality.
+
+There is not much to say about step 1, so let us explore steps 2, 3, and 4 in more detail.
+
+### Profiling
+In order to build our weighted CFG for layout analysis, we need to collect data about "control flow transfers" (when control transfers between CFG nodes). Luckily, most Intel CPUs come with a "last branch record", or LBR--- storing recently executed instructions such as jumps, branches, and calls. With the Linux `perf` tool, this data can be gathered and used to generate a weighted CFG for optimizing program layout.
+
+### Basic Block Chaining
+As we reorder blocks, we will need to insert additional jumps to preserve the control flow. Therefore, we wish to order basic blocks in a way that minimizes these unconditional jumps. This is to prevent unnecessary instruction pointer moves during the execution of the program. We define a basic block chain as a directed sequence of basic blocks where no unconditional jumps occur except possibly in the last block. Generating a maximum-cardinality set of these chains would constitute the fewest number of unconditional jumps required. The authors denote this problem the *fall-through maximization problem*. This problem is the same as the maximum path cover problem, which is **NP-hard**. PH uses a greedy heuristic to approach this but does not provide a theoretical guarantee of how close to optimal the generated solution is.
+
+There *is* is a 1/2-approximation algorithm for this problem guarantees that the weight of the path cover is within *1/2* of the optimal solution. While its performance does not absolutely exceed that of PH's heuristic, Codestitcher uses a hybrid of PH's method and the approximation algorithm, which provides better results than either algorithm run individually.
+
+### Hierarchical Code Collocation
+
+Coming up with a framework to integrate intraprocedural and interprocedural ordering optimizations while taking *layout distance* (the distance between two instructions) into account is difficult. We will first try to formalize this problem.
+
+#### *d*-Close Partial Layout
+
+It would help if we could formally define what layout distance really means. The paper's definition is quite intuitive: the number of bytes between two basic blocks. To take advantage of spatial locality, we want to ensure that related instructions that are executed frequently have as few bytes between them as possible. (This is because the instruction cache can only hold so many instructions.)
```

Comment:
FWIW, I think this is the central insight of the paper: that basic block layout should be based on minimizing the byte distance between pairs of blocks that are frequently executed together.

Reply:
> (This is because the instruction cache can only hold so many instructions.)

It is also because caches benefit from spatial locality because they're organized into cache lines. (That is, when you bring in one byte of a cache line, you get the rest of the line "for free," so it behooves you, as the compiler, to put frequently-used-together instruction within a given cache line. But you can't know for sure what the line size/alignment is, so it's almost as good to just generally put things as close together as you can.)